Support proxying multiple paths to different backends

md5 opened the pull request:
I'm finally opening a PR to review my branch for supporting multiple paths with `VIRTUAL_PATH`. I think that some of the duplicated code could probably be made a bit more DRY, but this code seems like a good starting point for a discussion to me.

A couple things are happening in this patch.
1. Detect if `VIRTUAL_PATH` is being used within the containers with a shared `VIRTUAL_HOST`
2. If so, take the SHA-1 hash of the path and use it as a suffix for both the `upstream` name and the `proxy_pass` statement

In the event that `VIRTUAL_PATH` is not set for `/` within the context of a particular `VIRTUAL_HOST`, a 404 is returned. This should possibly be a 503 instead, but I didn't want to hold up the discussion of this patch any longer (it's been over a year since I first created this branch).

Known issues that might need to be addressed before merging:
1. Further consolidate duplicate code into the shared "location" template
2. Figure out if 503 should be returned for a missing `/` instead of 404
3. Handle the case where some containers within a `VIRTUAL_HOST` have `VIRTUAL_PATH` set but others do not

Fixes #33

md5 commented:
Another thing that would need to happen if there is a desire to merge this PR is an update to `README.md` to document the feature.

md5 commented:
And I suppose some tests now that #246 has been merged. :+1:

md5 commented:
Looks like `groupBy` will leave out containers without the grouping key. I think a call to `whereNotExist` to find the containers without `VIRTUAL_PATH` may be needed, using `/` as the `.Path` for those containers.

md5 commented:
@jwilder I'd love to get your thoughts on this PR when you have a moment. Looks like the merge of #262 introduced a conflict, but I can fix that if you're interested.

jwilder commented:
@md5 What do you think about combining this functionality with #259 so that you can do all of this by just setting the `VIRTUAL_HOST` appropriately.

Something like:

```
VIRTUAL_HOST=example.com                  # works as is today
VIRTUAL_HOST=example.com/foo            # proxies example.com/foo to this container
VIRTUAL_HOST=example.com/foo:9090   # proxies example.com/foo to this container's port 9090
VIRTUAL_HOST=example.com,example.com/admin:8080   # proxies example.com to this container port 80 and example.com/admin to port 8080 on the container
```

There have been a few PRs for implementing parts of this functionality separately, but I think it would be easier for users if the syntax was combined.

Thoughts?

md5 commented:
I think the idea of combining `VIRTUAL_PATH` and `VIRTUAL_HOST` is a good one. I'm less keen on combining the port in that way since I think it could be confusing whether the port refers to the backend container port or the `nginx-proxy` port. There's also the fact that #259 includes setting a different protocol for different backends, which would further complicate the syntax for `VIRTUAL_HOST`.

At this point, I'd probably start moving toward a JSON config format, similar to what `ehazlett/interlock` does with `INTERLOCK_DATA`.

jwilder commented:
Yeah, originally I didn't like the idea of having the port part in there.  Starting to change my mind because the port for `nginx-proxy` is always what you define when you start the container.  I would be fine leaving that out initially though and still requiring `VIRTUAL_PORT` to change it.

I forgot about the protocol change piece to #259 as well.  I'm fine with keeping that out for now as well too.

md5 commented:
It looks like some work would need to be done in `docker-gen` to support the `VIRTUAL_HOST=host/some/path` syntax.

Namely, `docker-gen` would need to expose the `strings.SplitN` function to allow something like this:

 ```
{{ parts := splitN $host "/" 2 }}
{{ $host := $parts[0] }}
{{ $path := printf "/%s" $parts[1] }}
```

I'm not sure whether that's the actual code that would be needed; it seems like there could need to be a check to see if the `splitN` returned 1 or 2 parts and to act accordingly, but it may be that `text.template` allows indexing into undefined elements of a slice (unlikely).

md5 commented:
It just opened https://github.com/jwilder/docker-gen/pull/126 to add `splitN`. It appears that a `len` check will be necessary since `text.template` is strict about indexing into slices that are too short.

RheaUN commented:
Hello. I am not familiar with the netiquette within github, so I am not unsure if "+1" is welcome here. But I'll give it a try. :)

The feature would be really good in the way you propose. And it appears to be consistent/logical in the way the reverse-proxy is configured so far. 

The type of setup with a VIRTUAL_PATH can be very common in an environment with a DynDNS-Service in combination with a temporary IP-address.  Usually you have only one subdomain referring to your current IP-address/service. A VIRTUAL_PATH option would really help in this environment to reach different containers.

P.S.: THX for engagement in the reverse proxy!

md5 commented:
I've done some work toward getting this working with `splitN`.

One thing I realized is that there may be some differences of opinion as to how this feature should work. Specifically, I could see some users expecting the path to be stripped from the URL that is received before proxying it to the backend while others might want it to pass through unchanged.

As it stands now, the code will pass through the path unchanged. Whichever option is chosen, the backend application will need to know that it's running at `/path`, either to have its root URL be at `/path` or to know how to properly construct relative links to itself.

md5 commented:
We may want to pass an `X-Forwarded-Path` header to support informing the application which path it's running under.

jwilder commented:
I think setting `X-Forwarded-Path` is a good idea.  I also think the `/path` prefix should be stripped too so that the  backend is getting paths as if they were not mounted behind a proxy.

md5 commented:
Don't you think some users would want it one way and some the other way?

jwilder commented:
@md5 The use cases I've heard so far are to support exposing multiple backends where someone only has one domain name.  Making the path based proxying transparent to the backends seems simpler for users to me.  I.e they would be able to start there backends the same as if they were not using path based proxying (hopefully).

What do you think?

md5 commented:
I think that having a backend service that responds at `/` and wanting to expose it at `/path` is probably the more common use case, so it probably makes sense to do it first and have it be the default.

However, whichever option is chosen is not going to be transparent to the backend. Either way, the backend needs to know it's at `/path`. It either needs to have that be its root path for routing or it needs to know how to structure responses to be aware of the fact that relative paths are under `/path` instead of `/`.

md5 commented:
I'm taking a look again at supporting `VIRTUAL_HOST=host/path` again and I think this may be getting beyond the capabilities of `text.template`.

When I was using `VIRTUAL_HOST` and `VIRTUAL_PATH`, it was possible to group all the containers by `VIRTUAL_HOST`, then to group them by `VIRTUAL_PATH`. This allowed me to properly construct a `server` block per `VIRTUAL_HOST` with a nested `location` block per `VIRTUAL_PATH`.

In order to get the same structure with a single variable, I need to be able to group containers by the first part of the split `VIRTUAL_HOST` instead. The code would look something like this in Ruby:

```ruby
containers.reject { |c| c.Env['VIRTUAL_HOST'].empty? }
          .group_by { |c| c.Env['VIRTUAL_HOST'].split('/', 2)[0].trim }
          .each do |h,h_containers|

  puts "server { server_name #{h};"

  h_containers.group_by { |c| c.Env['VIRTUAL_HOST'].split('/', 2)[0].try(:trim) || '/' }
              .each do |p,_|

    upstream = "#{h}-#{Digest::SHA1.digest(p)}"
    puts "location #{p} { proxy_pass http://#{upstream}/; }"
  end

  puts "}"
end
```

Doing this with `docker-gen` would require adding something like a `groupBySplit` function, since it's not possible to define an anonymous function in `text.template` or to define a template function that can take another function as an argument.

If we didn't have to deal with the comma-separated list in `VIRTUAL_HOST`, I could see possibly passing a template name instead of a first class function like this:

```
{{define "virtualHost"}}{{ index (splitN (trim .Env.VIRTUAL_HOST) "/" 2) 0 }}{{end}}
{{ range $host, $containers := groupByTemplate $ "virtualHost" }}
```

However, I'm at a loss as to how to support the `groupByMulti` functionality used to split on "," as well. And now I realize that my example Ruby code above didn't deal with the comma case either...

md5 commented:
I've been thinking that it might be worth considering making `docker-gen` usable as a library instead of just as a command-line tool. That would allow an `nginx-proxy` binary to be built that has additional Nginx-specific functions beyond those provided by `docker-gen` itself. I think that could help clean up the template and make things like multiple path support easier to implement.

jwilder commented:
@md5 Yeah.  I've had the same thoughts recently and have implemented `docker-gen` like functionality a few times now for various things.  A custom `nginx-proxy` binary would make it easier to extend the project.

carlosbaraza commented:
Hi @md5 and @jwilder,

What is the current state of this pull request? I am very interested on the feature and so I'd be wiling to help, if needed.

By the way, my preferred usage would be to have backends responding at `/` and proxy mounting at `/path`.

Thank you guys for the work.

md5 commented:
@carlosbaraza The current state is that @jwilder asked me to try to support multiple paths by using an expanded syntax under `VIRTUAL_HOST` instead of adding a separate `VIRTUAL_PATH` environment variable. When I tried to implement that syntax, I found that it's not possible with the functions currently available for template expansion under `docker-gen`. Furthermore, I couldn't imagine a way to extend `docker-gen` to support the necessary functions without introducing functions that are highly specific to the `nginx-proxy` use case. There is also the additional problem that the `text/template` templating language is rather limited in terms of things like using first-class functions.

Since we had already been discussing the idea of making `docker-gen` usable as a library in addition to being usable as a binary command, the idea of making a binary `nginx-proxy` command that extends `docker-gen` was proposed. I started [a branch](https://github.com/appropriate/docker-gen/tree/generator) of `docker-gen` to make it usable as a library, but it doesn't currently compile and it has some more work to make the stock `docker-gen` binary usable as well as making it extensible for new use cases (in particular TLS configuration). Since I'm not really using `nginx-proxy` or `docker-gen` right now on a daily basis for my own work, I've had a hard time finding time to complete the work needed on the `generator` branch.

If you know something about Go programming and would like to work on making my `generator` branch usable for the stock `docker-gen` use case and extending it to producing customize binaries with additional functions, I'd be happy to review any PRs you can produce.

md5 commented:
An alternative would be to simply go ahead with an implementation that uses a separate `VIRTUAL_PATH` environment variable, since I already had that basically working

mgcrea commented:
A big :+1: for this feature, this would be tremendously useful. I've tried to patch things up myself but I've been struggling with the `text/template` capabilities. It would be really awesome if we could somehow abstract the templating engine. I'd love to be able to generate my configuration in a scripting language (like JS) from a plain JSON object.

mgcrea commented:
Is there anything that is missing is this PR? I have just tested the `nginx.tmpl` file from this PR and everything seems to work flawlessly.

Just had to add:
```
    {{ if (ne .Path "/") }}
    rewrite {{ .Path }}/(.*) /$1  break;
    {{ end }}
```
To get proper url rewriting.

RheaUN commented:
Dear All, with the chance to get bashed for a beginners question: I would like to implement the docker container with the VIRTUAL PATH function. What is he best way to create a docker container at my computer? I have tried to replace the changed files within the already created standard build, but the container didn’t start after that. I have checked https://guides.github.com/ but I am not sure how to actually fork this feature branch. I am using SourceTree.

What would be the best way to get this branch running on my machine? Dou you have a small clue form on what is the most reasonable next step to get  feature 254 working for me? THX!

md5 commented:
@mgcrea See my comment here: https://github.com/jwilder/nginx-proxy/pull/254#issuecomment-158845275

md5 commented:
@RheaUN The branch is at https://github.com/appropriate/nginx-proxy/tree/multiple-paths

If you already have a checkout of `jwilder/nginx-proxy`, I'd recommend adding my fork as another remote like so:

```console
$ git remote add appropriate git@github.com:appropriate/nginx-proxy.git
```

Then you can make your own branch from the `multiple-paths` branch like so:

```console
$ git checkout -b my-multiple-paths appropriate/multiple-paths
```

I haven't looked at this branch in a while, but it probably needs to have `master` merged into it from `jwilder/nginx-proxy` to make sure it's building on top of the latest functionality in the upstream repository.

mgcrea commented:
@RheaUN, an easy way would be to download the [raw nginx template](https://github.com/appropriate/nginx-proxy/blob/multiple-paths/nginx.tmpl) and mount it as a volume to the standard image. I'm currently using this `docker-compose.yml` file:

```yml
# export PROXY_IP=0.0.0.0.0

nginx_proxy:
  image: jwilder/nginx-proxy
  container_name: nginx_proxy
  hostname: nginx_proxy
  working_dir: /app
  environment:
    - DOCKER_HOST=unix:///tmp/docker.sock
  ports:
    - "${PROXY_IP}:443:443"
    - "${PROXY_IP}:80:80"
  volumes:
    - /var/run/docker.sock:/tmp/docker.sock:ro
    - /usr/bin/docker-gen:/usr/local/bin/docker-gen:ro
    - /opt/docker/nginx_proxy/certs:/etc/nginx/certs:ro
    - /opt/docker/nginx_proxy/nginx:/etc/nginx/conf.d
    - /opt/docker/nginx_proxy/docker-gen/nginx.tmpl:/app/nginx.tmpl
```

RheaUN commented:
Hello md5 and mgcrea, thanks for your help in showing a way to handle this! 
I took the time to go through my setup here trying to figure out what might be the best way to use the proxy.

I like the robustness and flexibility of the proxy, and both would be more or less reduced with a fork, as I experienced. 

I found a way, which works good in the dynamic IP setup (server @home with dynDNS service), which I would like to share, because it might help others with a similar setup and works around some limitations that usually appear in a non-commercial setup.

It is actually using VIRTUAL_HOST instead of VIRTUAL_PATH..., although you can usually link only one subdomain to your current dynamic IP:

- Update dynamic IP with dynDNS service to a subdomain _currentIP.domain.com_
- Edit the DNS configuration to add CNAME (canonical name) entry to point _subdomain1.domain.com_, _subdomain2.domain.com_, … to _currentIP.domain.com_

This works fine i.e. with all-incl.com as a domain / hosting provider, a standard router like the FritzBox and this reverse proxy.

Only “downside” I could see so far is that I’ll need SSL certificates for every subdomain. Quite obvious, but usually prevents the use in a non-commercial setup. This might be no issue with Let’s Encrypt in the future. 

Thanks again for the help and especially for the code!

sorinbadea-tremend commented:
Do you have an ETA for this PR into master?

md5 commented:
@sorinbadea-tremend My status on this is the same as before. I agree with @jwilder that it makes the most sense to combine the configuration into a single variable. His proposal was to use a special extended form for `VIRTUAL_HOST`, but I could see it making as much sense if not more to support an extended config via a JSON config string in an `NGINX_PROXY_CONFIG` variable.

In order to make this configuration work in a single environment variable, I'm pretty sure it would be necessary to have a custom `nginx-proxy` binary to allow custom functions specific to the Nginx use case to be implemented in Go. For that to be possible, `docker-gen` would need to be updated to separate its internals into a library and a standalone binary (see https://github.com/jwilder/docker-gen/issues/131 for discussion).

Due to an excess of other commitments and a lack of Go programming chops, I have not be able to successfully split `docker-gen` into a library and a binary in a way that I feel is maintainable. If someone else were able to muster the time and expertise to compete this task, I'm pretty confident I could add the necessary functions to a custom `nginx-proxy` binary based on this proposed `docker-gen` library to allow this multiple paths feature to be implemented with a single environment variable and Jason and I discussed.

jwilder commented:
I've pulled in @md5 start at converting docker-gen into a library and extended it a bit. Have not had time to work on it much lately though.

JanKoppe commented:
Hi, I'm very interested in this feature and eager to see it included. Is there anything that I could help to make it happen?

slavicandreev commented:
Would really love to see this feature. Any chance the merge is happening any time soon?

timofurrer commented:
Any chance to get this merged soon?

icecreammatt commented:
@timofurrer I found a nice alternative that supports this with similar container setup approach. https://traefik.io/ if you don't want to wait for this to be updated like myself.

qwIvan commented:
姿瓷

narfanator commented:
Jumping on the +1 bandwagon, since I'm trying to use this to dockerize the development environment of a NextJS app, and it seems like the only way for me to do what I want* is to use path mapping, since I have to use `localhost` as the host.

* I want to hand the artists (and other devs volunteering their time who're struggling to get things setup) a git repo with a three-step setup: 1) Install docker 2) `docker-compose up -d` 3) Visit `http://localhost`

Near as I can tell, I can only use this with an actual DNS name routed to some IP address.

gabadi commented:
+1

Brain-Gamer commented:
What is the current state of this pr? It would be sooooo nice not to have a subdomain for every container you have :D

md5 commented:
@Brain-Gamer I think the status is the same as [my last comment](https://github.com/jwilder/nginx-proxy/pull/254#issuecomment-183590213), but I haven't been keeping up with the development on `nginx-proxy` or `docker-gen`. It looks like some work was done to allow `docker-gen` to be used as a library in https://github.com/jwilder/docker-gen/pull/146, but I don't think any work has been done to make a purpose-specific `nginx-proxy` binary that uses it. I'm also not sure if the API of that library is sufficient to build an `nginx-proxy` binary that would be capable of doing everything that's currently done by calling `docker-gen` on the command line in addition to parsing the proposed configuration option(s) discussed earlier in this thread.

SatejG commented:
I need help in usage of "VIRTUAL_PATH".
Basically I want to setup "domain_name/app1", "domain_name/app2" for the different collection of containers.
I referred lots of discussion threads. But **I haven't got any final concrete solution over this**.
(
[1](https://github.com/appropriate/nginx-proxy/pull/1)
[2](https://github.com/jwilder/docker-gen/pull/47)
[3](https://github.com/jwilder/nginx-proxy/issues/190#issuecomment-117715206)
And more.......
)

Please guide in regards.

Sample docker-compose I used is given below:

------------------------------------------------------------------
version: '2'
services:
  # Create portainer container
  manage-docker:
    image: portainer/portainer:latest
    expose: ["9000"]
    container_name: portainer
    restart: always
    environment:
      - TERM=xterm
      - VIRTUAL_HOST=mydomain.test
      - VIRTUAL_PATH="/port-test"
    hostname: 'portainer.test'
    ports:
      - "9000:9000"         #default
    
    volumes:
      - /opt/my-data:/data
      - /var/run/docker.sock:/var/run/docker.sock

networks:
  default:
    external:
      name: my-network

oivindoh commented:
Any hope of merging this PR?

hamxabaig commented:
It works for me, I think its been merged.

CReimer commented:
Just stumbled upon this PR because I actually need this feature, too.
Could someone please pull this into the master branch. Thanks

andreycizov commented:
Any particular reason this is still not merged?

md5 commented:
@andreycizov The reasons are exhaustively covered in the previous comments.

m4teh commented:
Very much would like to see this also.

adrianilie9 commented:
Giving the fact that this is a very much requested feature that has been in stand-by for 2 years now because @jwilder considers the configuration template as not making "most sense" can we please get a pass and merge this?

It will only affect people whom are interested in this feature, which currently is non-existent. I think writing a couple more ENV variables trumps the lack of this feature.

md5 commented:
@adrianilie9 I haven't used this image in almost 3 years, so I'm not going to be working on this PR again. The only reason I haven't closed it is that I know people want the feature.

rodrigoaguilera commented:
I took another version of this feature and I rebased it against the current master in case some is interested

https://github.com/jwilder/nginx-proxy/pull/1083

raphaelyancey commented:
> I took another version of this feature and I rebased it against the current master in case some is interested
> 
> #1083

@rodrigoaguilera It works pretty well thanks a lot, but I was wondering why the path is passed to the target container as well? Would it be difficult to strip the virtual path off the requests before redirecting  to `VIRTUAL_HOST:VIRTUAL_PORT`?

Example: I'd like to serve a `phpmyadmin/phpmyadmin` container on `/phpmyadmin` but I can't since it requests `/phpmyadmin` to the target container, whereas it's served on `/`.

Or is there a nginx override I might create to make it working without additional code?

danieldaeschle commented:
any updates to this?

a1p4ca commented:
any updates to this?

SerdarN92 commented:
Any updates here? @md5 i really need this feature.

iedmrc commented:
Unbelievable! It's been 5 years since this pr is created.

adrianilie9 commented:
I have successfully ported this feature into a private fork. If anyone is still interested, i can make a public one. Just let me know.

iedmrc commented:
Why it hasn't been merge onto the master branch yet?

KrzysztofMadejski commented:
@adrianilie9, yes, please share it publicly and send as another PR. I'd love to use it as well. 

Did you resolve issues that md5 [mentioned](https://github.com/nginx-proxy/nginx-proxy/issues/190#issuecomment-117715206):

> The remaining open issues as I recall are:

> Sort the available VIRTUAL_PATH values for a VIRTUAL_HOST to ensure that dispatch will work in cases of prefix overlap (e.g. a container with VIRTUAL_PATH=/api alongside a container with VIRTUAL_PATH=/api/v1).

> Ensure that a 503 status code is still returned if VIRTUAL_PATH is being used and there is no container mapped to VIRTUAL_PATH=/.

KrzysztofMadejski commented:
I did catchup with laster master: https://github.com/KrzysztofMadejski/nginx-proxy/tree/multiple-paths

Try it with:
```
  whoami:
    image: jwilder/whoami
    environment:
      - VIRTUAL_HOST=whoami.local/site
```

kumarunster commented:
@adrianilie9 could you share your fork please?

camold commented:
+1 vote for obtaining a feature to specify different virtual paths/locations on the same host/server that are forwarded to different containers.
Would it not be possible to mount in a configuration file (as is done in section "Per-VIRTUAL_HOST location configuration" of the readme) for each virtual host and location combination? This may resolve the issue of configs (i.e. if the request path should also be forwarded to the container or not).
Concerning the ordering, would it be possible to use regular expressions? For instance "location = /" in the template? Then, the ordering should be irrelevant.

canhtran commented:
Hi Guys, how is thing PR going ? It really super helpful

N3ROO commented:
Any updates on this?

yoruvo commented:
Is this project abandoned? How come this PR hasn't been added in 6 years?

md5 commented:
> Is this project abandoned? How come this PR hasn't been added in 6 years?

The lack of updates to this PR should be quite clear from the comment history. I'm not sure there's anything more to say on that.

As for whether this project is abandoned, it looks like @jwilder made some updates in late 2020, but this is clearly not a project that is receiving feature updates. It looks like all recent PRs that were reviewed and merged are hotfixes to existing functionality.

buchdag commented:
This features has been added to the `dev` branch with #1607 

It's available now on the `dev` images `nginxproxy/nginx-proxy:dev` and `nginxproxy/nginx-proxy:dev-alpine`. Testing and feedback are welcome, the docs for the feature is [here](https://github.com/nginx-proxy/nginx-proxy/tree/dev#path-based-routing) and there is [another PR](https://github.com/nginx-proxy/nginx-proxy/pull/1699) opened for additional improvements and comments.
